src/ttl_table.c

```diff
@@ -0,0 +1,142 @@
+#include "ttl_table.h"
+#include "rmalloc.h"
+#include "util/minmax.h"
+#include "util/misc.h"
+
+typedef struct {
+  t_docId id;
+  t_expirationTimePoint point;
+} DocumentExpiration;
+
+typedef struct {
+  DocumentExpiration documentExpiration;
+  FieldExpiration* fieldExpirations;
+} TimeToLiveEntry;
+
+static uint64_t hashFunction_DocId(const void *key) {
+  return (uint64_t)key;
+}
+static void *dup_DocId(void *p, const void *key) {
+  return (void*)key;
+}
+static int compare_DocId(void *privdata, const void *key1, const void *key2) {
+  return (t_docId)key1 < (t_docId)key2;
+}
+static void destructor_DocId(void *privdata, void *key) {
+}
+
+static void destructor_TimeToLiveEntry(void *privdata, void *key) {
+    TimeToLiveEntry* entry = (TimeToLiveEntry*)key;
+    array_free(entry->fieldExpirations);
+    rm_free(entry);
+}
+
+static dictType dictTimeToLive = {
+  .hashFunction = hashFunction_DocId,
+  .keyDup = dup_DocId,
+  .valDup = NULL,
+  .keyCompare = compare_DocId,
+  .keyDestructor = destructor_DocId,
+  .valDestructor = destructor_TimeToLiveEntry,
+};
+
+void TimeToLiveTable_Init(TimeToLiveTable *table) {
+    if (!table->hashTable) {
+      table->hashTable = dictCreate(&dictTimeToLive, NULL);
+    }
+}
+
+void TimeToLiveTable_Destroy(TimeToLiveTable *table) {
+    if (table->hashTable) {
+      dictRelease(table->hashTable);
+      table->hashTable = NULL;
+    }
+}
+
+void TimeToLiveTable_Add(TimeToLiveTable *table, t_docId docId, t_expirationTimePoint docExpirationTime, arrayof(FieldExpiration) sortedById) {
+  TimeToLiveEntry* entry = (TimeToLiveEntry*)rm_malloc(sizeof(TimeToLiveEntry));
+  entry->documentExpiration.id = docId;
+  entry->documentExpiration.point = docExpirationTime;
+  entry->fieldExpirations = sortedById;
+  // we don't want the operation to fail so we use dictReplace
```

Comment:
The key is the document id.
Since we reindex the document on every change there will be a different document id every time.
So an update is not really possible in this case.

Reply:
so in what case can it fail? Sounds like we want at least to assert on the return value

Reply:
Okay, will change to dictAdd and add an assert